src/main/java/co/cask/hydrator/salesforce/SalesforceSchemaUtil.java

```diff
@@ -13,145 +13,89 @@
+      sObjectDescriptor.getAllParentObjects());
 
-    PartnerConnection partnerConnection = new PartnerConnection(connectorConfig);
-    DescribeSObjectResult describeSObjectResult = partnerConnection.describeSObject(sObjectName);
-
-    Map<String, Field> nameToField = new HashMap<>();
-    for (Field field : describeSObjectResult.getFields()) {
-      nameToField.put(field.getName(), field);
-    }
-    return nameToField;
+    return getSchemaWithFields(sObjectDescriptor, describeResult);
   }
 
   @VisibleForTesting
-  static Schema getSchemaWithFields(List<String> queryFields, Map<String, Field> nameToField) {
-
-    List<Schema.Field> schemaFields = new ArrayList<>(queryFields.size());
-    for (String fieldName : queryFields) {
-      Field salesforceField = nameToField.get(fieldName);
-      Schema fieldSchema;
-
-      if (isFieldComplex(fieldName)) {
-        // Some complex functions like nested and aggregate are not supported by Bulk API
-        // for relationship fields it would take too long to retrieve all fields for every sObject.
-        // Describe calls are a bit slow.
-        fieldSchema = Schema.of(Schema.Type.STRING);
-      } else {
-        if (salesforceField == null) {
-          throw new IllegalArgumentException(String.format("Field '%s' is absent the sObject", fieldName));
-        }
-
-        FieldType fieldType = salesforceField.getType();
-        fieldSchema = fieldTypeToCdapSchemaType(fieldType);
-
-        if (salesforceField.isNillable()) {
-          fieldSchema = Schema.nullableOf(fieldSchema);
-        } else {
-          fieldSchema = fieldSchema;
-        }
+  static Schema getSchemaWithFields(SObjectDescriptor sObjectDescriptor, SObjectsDescribeResult describeResult) {
+    List<Schema.Field> schemaFields = new ArrayList<>();
+
+    for (SObjectDescriptor.FieldDescriptor fieldDescriptor : sObjectDescriptor.getFields()) {
+      String parent = fieldDescriptor.hasParents() ? fieldDescriptor.getLastParent() : sObjectDescriptor.getName();
+      Field field = describeResult.getField(parent, fieldDescriptor.getName());
+      if (field == null) {
+        throw new IllegalArgumentException(
+          String.format("Field '%s' is absent in Salesforce describe result", fieldDescriptor.getFullName()));
       }
-
-      Schema.Field schemaField = Schema.Field.of(fieldName, fieldSchema);
+      Schema.Field schemaField = Schema.Field.of(fieldDescriptor.getFullName(), getCdapSchemaField(field));
       schemaFields.add(schemaField);
     }
 
     return Schema.recordOf("output",  schemaFields);
   }
 
-  /**
-   * If this is a relationship query (join equivalent for soql)
-   * or this is an aggregate function, or nested query as field,
-   * don't convert it from string to anything as this is very error-prone.
-   *
-   * Examples of complex fields:
-   * Relationship field - Account.Name
-   * Nested field - (SELECT Id FROM Opportunity)
```

Comment:
do we support nested fields correctly now?

Reply:
Ah never mind I see below that it throws an exception. Just want to confirm that in the past it would fail during query?

Reply:
Before it would fail during query execution, now it fails during query parsing.